src/Shared/ServerInfrastructure/StringUtilities.cs

```diff
@@ -17,102 +18,208 @@ internal class StringUtilities
         [MethodImpl(MethodImplOptions.AggressiveOptimization)]
         public static unsafe bool TryGetAsciiString(byte* input, char* output, int count)
         {
-            // Calculate end position
             var end = input + count;
-            // Start as valid
-            var isValid = true;
 
-            do
+            if (Sse2.IsSupported)
             {
-                // If Vector not-accelerated or remaining less than vector size
-                if (!Vector.IsHardwareAccelerated || input > end - Vector<sbyte>.Count)
+                if (Avx2.IsSupported && input <= end - Vector256<sbyte>.Count)
```

Comment:
pointers are unsigned and `end - Vector256<sbyte>.Count` can't go negative; so
```csharp
byte* p = null;
if (p <= p - 15)
{
    Console.WriteLine(true);
}
```
Will output `true`

So its better to switch the subtraction around with a `<=` test and have input overflow if it goes negative e.g.
```diff
- input <= end - Vector256<sbyte>.Count
+ input - Vector256<sbyte>.Count <= end
```

Reply:
Hm, this condition won't work.
E.g. this constellation is valid and could be processed by AVX-codepath:
```c#
byte* input = 0;
byte* end = 32;

if (input - Vector256<sbyte>.Count <= end)
{
    // ...
}
```
Here it will wrap around to `0xffffffffffffffe0 <= 32` which is false.

So it should be `input + Vector256<sbyte>.Count <= end` which itself can overflow, hence the condition is written with the subtraction.

Your concern is valid, so I could add
```diff
public static unsafe bool TryGetAsciiString(byte* input, char* output, int count)
{
+   Debug.Assert(input != null);
+   Debug.Assert(output != null);
```
?

Reply:
Or just `count >= Vector256<sbyte>.Count`?

Reply:
That's the easy way 😃 

In https://github.com/dotnet/aspnetcore/blob/80a77ff5c23647fa52782d15d0abb92cd2afbbb1/src/Shared/ServerInfrastructure/StringUtilities.cs#L49 the same conditions is used, the JIT will use the value from a register. I'll need to check codegen / check if the loop condition in still kept in a register.

And in the lines below -- either for differnt code-path or to process the remainder -- similar conditions are used. To use `count` here this value needs bookkeeping. 

As this is a `Try...`-method I think it's better to just `return false` in case `input == null`.

Reply:
Is it provably impossible for `end < Vector256<sbyte>.Count` if we know it's not null? I know we were already doing `end - Vector<sbyte>.Count` before, but I'm curious.

Reply:
In theory `end < Vector256<sbyte>.Count` can be `true`, but not in practice. 
Even if `count = 0` then `end` will be the same as `input` and that pointer won't reside at addresses lower than `32`. Hence there is no real problem. 
As it's internal I'll add another assert for this.